Add authentication doc (#1)
shirhatti/dotnet-monitor: documentation/authentication.md

> Authentication

Reviewer: I would suggest replacing all references to "API Token" with "API Key" since we aren't using tokens.

> ================

Reviewer 2: ```suggestion
# Authentication
```

nit: Use consistent markdown syntax

> Authenticating requests to `dotnet-monitor` helps protect sensitive diagnostic artifacts from unauthorized users and lower privileged processes. `dotnet-monitor` can be configured to use either [Windows Authentication](#windows-authentication) or via an [API Token](#api-token-authentication). It is possible, although strongly not recommended, to [disable authentication](#disabling-authentication).

Reviewer: ```suggestion
Authenticated requests to `dotnet-monitor` help protect sensitive diagnostic artifacts from unauthorized users and lower privileged processes. `dotnet-monitor` can be configured to use either [Windows Authentication](#windows-authentication) or via an [API Key](#api-key-authentication). It is possible, although strongly not recommended, to [disable authentication](#disabling-authentication).
```

> Authenticating requests to `dotnet-monitor` helps protect sensitive diagnostic artifacts from unauthorized users and lower privileged processes. `dotnet-monitor` can be configured to use either [Windows Authentication](#windows-authentication) or via an [API Token](#api-token-authentication). It is possible, although strongly not recommended, to [disable authentication](#disabling-authentication).

Reviewer 2: We should add something here to mention we only do authentication on a secure channel and link to the https doc (which I realized doesn't exist in the TOC)

Reviewer 2: We should also mention we only perform authentication on non-metrics endpoints

Author: Is it true that we only authenticate on a secure channel? I mentioned this down on Line 14, but I believe the behavior is to emit a warning when TLS is not enabled. @wiktork / @jander-msft - could one of you confirm?

Reviewer 2: If not, let's add something in documentation to discourage using authentication over an insecure channel.

Reviewer: We do allow authentication through plain http, but we log a warning message when configured that way:

`WARNING: Authentication is enabled over insecure http transport. This can pose a security risk and is not intended for production environments.`

## Windows Authentication

> We only recommend using Windows Authentication if you're running `dotnet-monitor` as a local development tool on Windows, for all other environments using an [API Token](#api-token-authentication) is recommended.

Reviewer: nit:
```suggestion
We only recommend using Windows Authentication if you're running `dotnet-monitor` as a local development tool on Windows; for all other environments using an [API Token](#api-token-authentication) is recommended.
```

> Windows authentication doesn't require explicit configuration and is enabled automatically when running `dotnet-monitor` on Windows. When available, dotnet-monitor will authorize any user authenticated as the same user that started the `dotnet-monitor` process. It is not possible to disable Windows authentication.

Reviewer 3: Not sure if it's mentioned, but we will not attempt windows auth if you run dotnet-monitor as administrator

> **FACT CHECK:** Windows Authentication cannot be disabled when running on Windows. See: [DiagnosticsMonitorCommandHandler](https://github.com/dotnet/dotnet-monitor/blob/69542c44b7d70a83901b93a73d0ea09b90870bcd/src/Tools/dotnet-monitor/DiagnosticsMonitorCommandHandler.cs#L136)

## API Token Authentication
An API Token is the recommended authentication mechanism for `dotnet-monitor`. To enable it, you will need to generate a secret key, update the configuration of `dotnet monitor`, and then specify the API Token in the Authorization header on all requests to `dotnet-monitor`.

> **NOTE:** API Token Authentication should only be used when [TLS is enabled](#) to protect the token while in transit.

### Generating an API Token

> The API Token you use to secure `dotnet-monitor` should be a 32-bit cryptographically random secret. You can generate a token either using `dotnet-monitor` or via your shell. To generate an API Token with `dotnet-monitor`, simply invoke the `generatekey` command:

Reviewer: It's 32 bytes (256 bits) not 32 bits.

```powershell
dotnet-monitor generatekey
```

The output from this command will display the API Token formatted as an authentication header along with its hash and associated hashing algorithm. You will need to store the `ApiKeyHash` and `ApiKeyHashType` in the configuration for `dotnet-monitor` and use the authorication header value when making requests to the `dotnet-monitor` HTTPS endpoint.

```
Authorization: MonitorApiKey Uf6Yq8ZGkcn+ltq2QLHxuXpKA6/Q5VH5Mb5aSIJBxRc=
ApiKeyHash: CB233C3BE9F650146CFCA81D7AA608E3A3865D7313016DFA02DAF82A2505C683
ApiKeyHashType: SHA256
```

> > NOTE: The generated API token should be secured at rest. We recommend using a tool such as a password manager to save it.

Reviewer 3: Note that other hash algorithms are allowed, but md5 and sha1 are not.

Author: Happen to have handy the list of hash algorithms that *are* allowed, @wiktork?

Reviewer: https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.hashalgorithm.create?view=net-5.0#System_Security_Cryptography_HashAlgorithm_Create_System_String_

Effectively, anything in that list that is not SHA1 or MD5 (basically SHA256, SHA384, and SHA512).

If you're running on Windows, you can save these settings to `%USERPROFILE%\.dotnet-monitor\settings.json`. If you're on other operating systems, you can save this file to `$XDG_CONFIG_HOME/dotnet-monitor/settings.json`.

> **NOTE:** If `$XDG_CONFIG_HOME` isn't defined, dotnet-monitor will fall back to looking at `$HOME/.config/dotnet-monitor/settings.json`

```json
{
  "ApiAuthentication": {
    "ApiKeyHash": "<HASHED-TOKEN>",
    "ApiKeyHashType": "SHA256"
  }
}
```

Alternatively, you can use environment variables to specify the configuration.

```sh
DotnetMonitor_ApiAuthentication__ApiKeyHash="<HASHED-TOKEN>"
DotnetMonitor_ApiAuthentication__ApiKeyHashType="SHA256"
```

> **NOTE:** When you use environment variables to configure the API Key hash, you must restart `dotnet-monitor` for the changes to take effect.

### Configuring an API Token in a Kubernetes Cluster
If you're running in Kubernetes, we recommend creating secrets and mounting them into container via a deployment manifest.

```sh

> kubectl create secret generic apikey \

Reviewer 3: note this command can also be used to do a key rotation. Not sure you want to mention that only secrets mounted as volumes will support key rotations without restarts.

  containers:
  - name: dotnetmonitoragent
    image: mcr.microsoft.com/dotnet/dotnet-monitor:5.0.0-preview.4
    volumeMounts:
      - name: apikey
        mountPath: /etc/dotnet-monitor

```

> **NOTE:** For a complete example of running dotnet-monitor in Kubernetes, see [Running in a Kubernetes Cluster](getting-started.md#running-in-a-kubernetes-cluster) in the Getting Started guide.

### Generate an API Token with PowerShell
```powershell
$rng = New-Object System.Security.Cryptography.RNGCryptoServiceProvider
$secret = [byte[]]::new(32)
$rng.GetBytes($secret)
$API_TOKEN = [Convert]::ToBase64String($secret)
"Authorization: MonitorApiKey $API_TOKEN"

$secretStream = [System.IO.MemoryStream]::new($secret)
$HASHED_TOKEN = (Get-FileHash -Algorithm SHA256 -InputStream $secretStream).Hash
$rng.Dispose()

Write-Output "ApiKeyHash: $HASHED_TOKEN"
Write-Output "ApiKeyHashType: SHA256"
```

### Generate an API Token with a Linux shell

```sh
API_TOKEN=`openssl rand -base64 32`
echo "Authorization: MonitorApiKey $API_TOKEN"

HASHED_TOKEN=`$API_TOKEN | base64 -d | sha256sum`
echo "ApiKeyHash: $HASHED_TOKEN"
echo "ApiKeyHashType: SHA256"
```

## Authenticating requests

> When using Windows Authentication, your browser with automatically handle the Windows authentication challenge. If you are using an API Token, you must specify it via the authentication header.

Reviewer: ```suggestion
When using Windows Authentication, your browser will automatically handle the Windows authentication challenge. If you are using an API key, you must specify it via the authentication header.
```

```sh

> curl -H "Authorization: MonitorApiKey HdFb8OPkE0Dc5hpu0kAxA7hhNguAah9SNUFftlP2Dk0=" https://localhost:52323/processes

Reviewer: Could you use the same encoded key throughout the example: `Uf6Yq8ZGkcn+ltq2QLHxuXpKA6/Q5VH5Mb5aSIJBxRc=`

Reviewer 2: I'd like to see the Windows version of this example as well using `curl.exe`. At least in Powershell, `curl` is an alias for `Invoke-WebRequest` that doesn't accept the same parameters

> **NOTE:** If `$XDG_CONFIG_HOME` isn't defined, dotnet-monitor will fall back to looking at `$HOME/.config/dotnet-monitor/settings.json`

```json
{
  "ApiAuthentication": {
    "ApiKeyHash": "CB233C3BE9F650146CFCA81D7AA608E3A3865D7313016DFA02DAF82A2505C683",
    "ApiKeyHashType": "SHA256"
  }
}
```

Alternatively, you can use environment variables to specify the configuration.

```sh
DotnetMonitor_ApiAuthentication__ApiKeyHash="CB233C3BE9F650146CFCA81D7AA608E3A3865D7313016DFA02DAF82A2505C683"
DotnetMonitor_ApiAuthentication__ApiKeyHashType="SHA256"
```

> **NOTE:** When you use environment variables to configure the API Key hash, you must restart `dotnet-monitor` for the changes to take effect.

### Configuring an API Key in a Kubernetes Cluster
If you're running in Kubernetes, we recommend creating secrets and mounting them into container via a deployment manifest. You can use this `kubectl` command to either create or rotate the API Key.

```sh
kubectl create secret generic apikey \
	--from-literal=ApiAuthentication__ApiKeyHash=$hash \

> 	--dry-run=client -o yaml \

Reviewer 3: I would add the --from-literal=ApiAuthentication__ApiKeyHashType=SHA256 line here too. K8 will mount both settings into the file system.

Reviewer 3: Related to this is https://github.com/dotnet/dotnet-monitor/issues/164. We don't make it easy to mount secret and non-secret data from multiple sources.

"Authorization: MonitorApiKey $API_KEY"

$secretStream = [System.IO.MemoryStream]::new($secret)
$HASHED_KEY = (Get-FileHash -Algorithm SHA256 -InputStream $secretStream).Hash
$rng.Dispose()

Write-Output "ApiKeyHash: $HASHED_KEY"
Write-Output "ApiKeyHashType: SHA256"
```

### Generate an API Key with a Linux shell

```sh
API_KEY=`openssl rand -base64 32`
echo "Authorization: MonitorApiKey $API_KEY"

HASHED_KEY=`$API_KEY | base64 -d | sha256sum`
echo "ApiKeyHash: $HASHED_KEY"
echo "ApiKeyHashType: SHA256"
```

## Authenticating requests
When using Windows Authentication, your browser will automatically handle the Windows authentication challenge. If you are using an API Key, you must specify it via the authentication header.

```sh
curl -H "Authorization: MonitorApiKey Uf6Yq8ZGkcn+ltq2QLHxuXpKA6/Q5VH5Mb5aSIJBxRc=" https://localhost:52323/processes
```

If using PowerShell, be sure to use `curl.exe`, as `curl` is an alias for `Invoke-WebRequest` that does not accept the same parameters.

```powershell
curl.exe -H "Authorization: MonitorApiKey Uf6Yq8ZGkcn+ltq2QLHxuXpKA6/Q5VH5Mb5aSIJBxRc=" https://localhost:52323/processes
```

## Disabling Authentication

> Disabling authentication could enable lower privileged processes to exfiltrate sensitive information, such as the full contents of memory, from any .NET Core application running within the same boundary. You should only disable authentication when you have evaluated and mitigated the security implications of running `dotnet-monitor` unauthenticated.

Reviewer 2: ```suggestion
Disabling authentication could enable lower privileged processes to exfiltrate sensitive information, such as the full contents of memory, from any .NET application running within the same boundary. You should only disable authentication when you have evaluated and mitigated the security implications of running `dotnet-monitor` unauthenticated.
```

nit: Drop Core to avoid any .NET 5/6 confusion
